### PR TITLE
Use the archive file name format for the compressed file when using compression

### PR DESCRIPTION
**Pull request by TalAloni**
This PR sets the entry name for the filename inside the zip (when compression is enabled) according to the archiveFileName format.

**Motivation:** I have recently enabled compression and found myself in a situation where I cannot mass-extract archived log files to the same directory because they are all using the same filename inside the archive. it seems to me that the rules for the archive filename should also apply to the entry name inside the archive.
Simply put, I expect that once I extract the zip files, I will have the same filenames that I had before enabling compression.

**Additional information:** 
I have worked with the following configuration
```
<target name="file" xsi:type="File" fileName="${basedir}/logs/log-latest.csv" archiveFileName="${basedir}/logs/log.{#####}.csv">
```
Now I wish to enable compression using the following configuration:
```
<target name="file" xsi:type="File" fileName="${basedir}/logs/log-latest.csv" archiveFileName="${basedir}/logs/log.{#####}.zip" enableArchiveFileCompression="true">
```

### CODE REVIEW

**Comment by snakefoot** [comment]
This change makes a lot of sense, but is ofcourse a breaking change, and will have to wait for NLog 5.0

I guess the FileTarget can eat one more bullet, but I think it would be nice if the FileTarget was stripped from any copy/move/delete/compress-logic. Instead the file-manipulation-logic should be implemented using one (or more) FileTargetWrapper. See also #2732 and  #1249

**Comment by snakefoot** [comment]
Closing and reopening to rebuild on Travis.

**Comment by snakefoot** [comment]
Besides being a behavior breaking change, then it also breaks the `IFileCompressor`-interface.

I guess very few have made override of the `IFileCompressor`-interface, but would prefer if the upgrade-path was less breaking. Thus introducing a new interface, and make a wrapper for the old interface-property and mark the old interface as obsolete.

**Comment by TalAloni** [comment]
Thanks,
There's already a FileCompressor property of type IFileCompressor.
It seems to me you're suggesting we add an "IArchiveFileCompressor" that inherits from IFileCompressor, and then check if we're dealing with IArchiveFileCompressor or the now obsolete IFileCompressor.

This seems a bit excessive, but I'll do it if you think it's necessary.
IMO given that not many inherited the interface maybe it's best to just reduce noise - it will be simple enough for implementing classes to add the new parameter to the implementation - given that it is targeted for a major version I think that's acceptable.
I don't think that it will be clearly understood that you can implement IArchiveFileCompressor if the property type is IFileCompressor - so I find the approach you're describing less appealing.

**Comment by snakefoot** [comment]
I was thinking 3 items:
- IArchiveFileCompressor - New interface
- LegacyArchiveFileCompressor - Implements new interface and takes legacy IFileCompressor as input-parameter
- ZipArchiveFileCompressor - Implements new interface

The do the following:
- Mark IFileCompressor as obsolete
- Mark FileTarget FileCompressor-property as obsolete
- Change FileTarget FileCompressor-property implementation. Let the setter allocate LegacyArchiveFileCompressor and give the propery-value as constructor-parameter. Let the getter return the IFileCompressor inside the LegacyArchiveFileCompressor.

**Comment by TalAloni** [comment]
I see,
We can also just change the implementation of ZipArchiveFileCompressor - so it would contain the `Path.GetFileNameWithoutExtension(archiveFileName) + Path.GetExtension(fileName)`
logic - this way the interface will remain the same - if that's important to you than maybe we should settle for this simple solution.
Let me know what you think is best. thanks!

**Comment by snakefoot** [comment]
Like that work-around. Replacing interface and guiding users to the new-interface generates a lot of code-bloat.

**Comment by TalAloni** [comment]
Great, I've made the change - the interface remain unchanged. this is tested and working as expected.

**Comment by snakefoot** [comment]
Very elegant solution. I like it.

Tried to figure out whether there was any special reasons for the initial solution (with static filename). But it seems it was just hacked together until it worked, without much consideration about entry-filename. But still breaking change so waiting for NLog 5.0

**Comment by snakefoot** [comment]
@TalAloni Could you extend one of the existing unit-test to verify that the entry-filename is as expected? (Then you will also have better gurantee that it doesn't break in the future).

**Comment by snakefoot** [comment]
@TalAloniI would prefer that `AssertZipFileContents` was used for the unit-test to verify the contents of final zip-file (Having the expected the Entry-Name)

**Comment by TalAloni** [comment]
This would mean adding an additional parameter to AssertZipFileContents (archiveFileName or entryName) and update a few dozens existing tests.
Instead we can add a new AssertZipFileEntryName which would be less obtrusive, which do you prefer?

**Comment by snakefoot** [comment]
Was thinking you could do the exact same trick you did in `ZipArchiveFileCompressor.cs` to extract the expected Entry-Name from the input-filename. (Where EntryName has `.txt`-file-extension).

**Comment by TalAloni** [comment]
Currently AssertZipFileContents only gets the zip file name - it does not have a way to check if the entry name (including the extension) inside the zip match the expected entry name without an additional parameter (the expected entry name or the log file name before archiving).
All we can do without adding an additional parameter is to check the entry name without the extension.

**Comment by snakefoot** [comment]
> All we can do without adding an additional parameter is to check the entry name without the extension.

Pretty sure all test-cases use `.txt` as file-extension, so just go with that as hardcoded value. When someone needs a test with a different file-extension, then it can be improved.

**Comment by snakefoot** [comment]
Curious with this failing test:

```
      Assert.Equal() Failure
      Expected: archive.txt
      Actual:   tmpE9E9.tmp
      Stack Trace:
        tests\NLog.UnitTests\NLogTestBase.cs(189,0): at NLog.UnitTests.NLogTestBase.AssertZipFileContents(String fileName, String contents, Encoding encoding)
        tests\NLog.UnitTests\Targets\FileTargetTests.cs(796,0): at NLog.UnitTests.Targets.FileTargetTests.ArchiveFileOnStartTests(Boolean enableCompression, Boolean customFileCompressor)
```

**Comment by snakefoot** [comment]
Also looks like the rolling-archive-mode is not working well with this feature. When "rolling" the zip-files (renames it) then it doesn't roll the entry-name inside the zip-file. Guess the filename should remain static (the old way) when `ArchiveNumbering.Rolling`

**Comment by TalAloni** [comment]
Unfortunately we no longer have access to ArchiveNumbering since we moved the logic to the ZipArchiveFileCompressor in order to avoid changing the interface.

**Comment by TalAloni** [comment]
I can easily fix the failing test by appending .txt to Path.GetTempFileName()

**Comment by TalAloni** [comment]
I think the original PR made more sense, especially given ArchiveNumbering.Rolling
and I also think it's OK to expand interfaces when releasing a major version, especially ones that are rarely implemented

**Comment by snakefoot** [comment]
> I can easily fix the failing test by appending .txt to Path.GetTempFileName()

That will be good start.

> Unfortunately we no longer have access to ArchiveNumbering since we moved the logic to the ZipArchiveFileCompresso

Maybe something like this:

```c#
        public ArchiveNumberingMode ArchiveNumbering
        {
            get => _archiveNumbering;
            set
            {
                if (_archiveNumbering != value)
                {
                    _archiveNumbering = value;
                    ResetFileAppenders("ArchiveNumbering Changed"); // Reset archive file-monitoring
                    (FileCompressor as ZipArchiveFileCompressor)?.SetArchiveNumbering(value);
                }
            }
        }
        private ArchiveNumberingMode _archiveNumbering;
```

**Comment by TalAloni** [comment]
The proposed change to ZipArchiveFileCompressor feels counter productive, as this would only be available to the internal ZipArchiveFileCompressor while IFileCompressor would be less flexible.
I stand by my original PR - I think that it's more appropriate to have the entry name come from the consumer of the interface.
I think that you put too much emphasis on maintaining the interface signature and you shouldn't let that from interfering with the right design choice - ask yourself how the interface would have looked like if you would have started from scratch, and that's what the interface should become. it would be best to avoid making one odd workaround after another.

**Comment by snakefoot** [comment]
> and I also think it's OK to expand interfaces when releasing a major version,

Think my goal is to remove a lot of the logic from FileTarget, and instead have some lifetime events, that a wrapper can react to. Ex. archive-event that triggers compression in the background, instead of blocking the application. Or a wrapper that wraps newly create FileStream-objects so they actually creates a gzip-stream. Etc.

Would also like to split FileTarget into two, so it becomes a vanilla FileTarget without ConcurrentWrites-support (Thus no FileWatcher or global mutex, etc.). And a ConcurrentFileTarget in a seperate nuget-package.

Like to post-pone breaking changes until they are really needed or gives a huge benefit. And if possible provide an easy migration path without too many runtime-errors.

**Comment by snakefoot** [comment]
> it would be best to avoid making one odd workaround after another.

The "problem" right now is that an unit-test has revealed that the pull-request is not working well with `ArchiveNumbering.Rolling`. You are ofcourse welcome to go back to my [original suggestion](https://github.com/NLog/NLog/pull/4189#issuecomment-761881813). But right now I prefer the path of just making `ZipArchiveFileCompressor` work, and if it needs a hint, then I'm okay with that.

**Comment by TalAloni** [comment]
OK, I added a hint to ZipArchiveFileCompressor like you prefer,
AssertZipFileContents will need an additional parameter take into account ArchiveNumberingMode. this means updating a few dozens tests.

**Comment by snakefoot** [comment]
Not sure how 3 locations turns into dozens, but yes having the ability to disable entry-name validation for `RollingArchiveTests` should be enough.

File: src/NLog/Targets/ZipArchiveFileCompressor.cs [hunk 1]
```diff
@@ -50,17 +50,24 @@ internal class ZipArchiveFileCompressor : IFileCompressor
         /// </summary>
         public void CompressFile(string fileName, string archiveFileName)
         {
+            string entryName = IsEntryNameInferredFromArchiveFileName ? (Path.GetFileNameWithoutExtension(archiveFileName) + Path.GetExtension(fileName)) : fileName;
```
Comment (snakefoot) [inline src/NLog/Targets/ZipArchiveFileCompressor.cs:53]: Think it should be (Missing `Path.GetFileName`):
`string entryName = IsEntryNameInferredFromArchiveFileName ? (Path.GetFileNameWithoutExtension(archiveFileName) + Path.GetExtension(fileName)) : Path.GetFileName(fileName);`

Comment (TalAloni) [reply src/NLog/Targets/ZipArchiveFileCompressor.cs]: Of course - thank you

**Comment by TalAloni** [comment]
Sorry, I miscounted the usages of AssertZipFileEntryName.
I have implemented the change in the way you preferred, all tests pass.

**Comment by snakefoot** [comment]
Very happy with the updated unit tests. The unit tests for the FileTarget are rather gnarly. So thank you for your persistence on making them green again with the extra validation.


Sent from my Sony Xperia


---- TalAloni wrote ----


Sorry, I miscounted the usages of AssertZipFileEntryName.
I have implemented the change in the way you preferred, all tests pass.

**Review by snakefoot — APPROVED** [review]

**Comment by snakefoot** [comment]
I can now see that I have given wrong advice. The FileCompressor-property is a static property, so having a single FileTarget changing properties on this static-instance will not work well, when using multiple FileTargets in the same NLog.config.

So the entry-name must be given as input-parameter when calling the IFileCompressor. To provide safe passage, then a V2-interface must be introduced (that includes the method with entry-name-parameter). And then call this people when the static-instance implements the V2-interface (else call the old interface-method).

**Comment by TalAloni** [comment]
"So the entry-name must be given as input-parameter when calling the IFileCompressor"
Yes, this approach is healthier.

"To provide safe passage, then a V2-interface must be introduced"
Again, I think it's too cumbersome, I think it's extremely rare for users to implement this interface, we add a lot of noise for background compatibility that nobody really care for, and that could easily be updated to the new interface.
I will do it, but only to please you, I think that it's a messy and confusing approach. using the existing interface and property would be much cleaner and easier for most users, if not all.

**Comment by snakefoot** [comment]
This will also do:

```c#
if (FileCompressor is ZipArchiveFileCompressor defaultCompressor)
   defaultCompressor.CompressFile(fileName, archiveFileName, entryFileName);
else 
   FileCompressor.CompressFile(fileName, archiveFileName);
```

**Comment by TalAloni** [comment]
I've made the requested changes. I've added a new interface extending the existing interface to improve extendibility.

**Comment by TalAloni** [comment]
Thanks,
The idea beyond making IArchiveFileCompressor public is to allow developers to set FileCompressor to an IArchiveFileCompressor implementation, this way they be able to have access to the 'entryName' argument.

**Comment by snakefoot** [comment]
The NLog FileTarget has so many door-knobs that it is difficult for normal users to figure out how to use them. Several of the door-knobs are counter intuitive, and make normal users think the NLog FileTarget is broken. Adding more public door-knobs will not improve the situation.

Like I said initially "I guess the FileTarget can eat one more bullet, but I think it would be nice if the FileTarget was stripped" (down).

**Comment by snakefoot** [comment]
Sorry for taking so long to get this pull-request merged. But very grateful for the unit-tests, they have been missing since the zip-feature was introduced.

Maybe one day a champion will arrive that will carve the NLog FileTarget into smaller pieces, so it will be easier to maintain and with improved platform compability.

**Comment by TalAloni** [comment]
Thanks for the great and very useful project!

**Comment by snakefoot** [comment]
Updated the wiki: https://github.com/NLog/NLog/wiki/File-target

**Pull request merged**